File: python/beeai_framework/agents/ability/abilities/handoff.py
from functools import cached_property

from pydantic import BaseModel, Field

from beeai_framework.agents.ability.abilities.ability import Ability, AbilityCheckResult, with_run_context
from beeai_framework.agents.ability.agent import AbilityAgent
from beeai_framework.agents.ability.types import AbilityAgentRunState
from beeai_framework.context import RunContext
from beeai_framework.memory import BaseMemory


class HandoffSchema(BaseModel):
    prompt: str = Field(description="Clearly defined task for the agent to work on based on his abilities.")


class HandoffAbility(Ability[HandoffSchema]):
    """Delegates a task to an expert agent"""

    def __init__(self, target: "AbilityAgent", *, name: str | None = None, description: str | None = None) -> None:
        self._target = target

        self.name = name or target.meta.name
        self.description = description or target.meta.description
        # self.description += "(context must contain only verified information)"

        super().__init__()

    @cached_property
    def input_schema(self) -> type[HandoffSchema]:
        return HandoffSchema

    def check(self, states: AbilityAgentRunState) -> AbilityCheckResult:
        return AbilityCheckResult(allowed=True, forced=False, hidden=False, prevent_stop=False)

    @with_run_context
    async def run(self, obj: HandoffSchema, context: RunContext) -> str:
        memory: BaseMemory = context.context["state"]["memory"]

        target = await self._target.clone()
        await target.memory.add_many(memory.messages[1:])
        response = await target.run(prompt=obj.prompt)
        return response.result.text

File: typescript/src/tools/handoff.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { z } from "zod";
import {
  BaseToolRunOptions,
  StringToolOutput,
  Tool,
  ToolEmitter,
  ToolError,
  ToolInput,
} from "./base.js";
import { BaseMemory } from "@/memory/base.js";
import { AssistantMessage, Message, SystemMessage, UserMessage } from "@/backend/message.js";
import { Emitter } from "@/emitter/emitter.js";
import { AnyAgent } from "@/agents/types.js";
import { GetRunContext } from "@/context.js";
import { getProp } from "@/internals/helpers/object.js";
import { findLastIndex, toCamelCase } from "remeda";

export interface HandoffToolInput {
  name?: string;
  description?: string;
  propagateInputs?: boolean;
}

export class HandoffTool extends Tool<StringToolOutput> {
  private readonly propagateInputs: boolean;
  public name: string;
  public description: string;

  public readonly emitter: ToolEmitter<ToolInput<this>, StringToolOutput> = Emitter.root.child({
    namespace: ["tool", "handoff"],
    creator: this,
  });

  inputSchema() {
    return z.object({
      task: z
        .string()
        .describe("Clearly defined task for the agent to work on based on his abilities."),
    });
  }

  constructor(
    private target: AnyAgent,
    options?: HandoffToolInput,
  ) {
    super();
    this.name = toCamelCase(options?.name || this.target.meta.name);
    this.description = options?.description || this.target.meta.description;
    this.propagateInputs = options?.propagateInputs ?? true;
  }

  protected async _run(
    input: ToolInput<this>,
    options: Partial<BaseToolRunOptions>,
    run: GetRunContext<typeof this>,
  ): Promise<StringToolOutput> {
    const memory = getProp(run.context, [Tool.contextKeys.Memory]) as BaseMemory;

    if (!memory) {
      throw new ToolError(`No memory found in the context.`, [], {
        isRetryable: false,
        isFatal: true,
      });
    }

    let messages: Message[] = memory.messages.filter((msg) => !(msg instanceof SystemMessage));

    const lastValidMsgIndex = findLastIndex(
      messages,
      (msg) => msg instanceof AssistantMessage && (msg.getToolCalls()?.length ?? 0) > 0,
    );

    if (lastValidMsgIndex !== -1) {
      messages = messages.slice(0, lastValidMsgIndex);
    }

    if (this.propagateInputs) {
      messages.push(new UserMessage(input.task));
    }

    const target = await this.target.clone();
    target.memory.reset();
    await target.memory.addMany(messages);

    const response = await target.run({ prompt: null });

    return new StringToolOutput(response.result.text);
  }
}
